File: src/core/repetition_detector.py
import re

class RepetitionDetector:
    def __init__(self, n=20, max_repeats=5):
        self.n = n
        self.max_repeats = max_repeats
        from collections import deque
        self.buffer = deque(maxlen=n * max_repeats)

    def add_text(self, text):
        if text is None: return True
        tokens = re.findall(r'\w+', text)
        
        for token in tokens:
            self.buffer.append(token)
            if len(self.buffer) < self.n * 2: continue
            
            # Check if the last n tokens match the previous n tokens
            current_ngram = tuple(list(self.buffer)[-self.n:])
            previous_ngram = tuple(list(self.buffer)[-self.n*2:-self.n])
            
            if current_ngram == previous_ngram:
                # Simple count of occurrences in the fixed-size buffer
                count = 0
                buf_list = list(self.buffer)
                for i in range(0, len(buf_list) - self.n + 1, self.n):
                    if tuple(buf_list[i:i+self.n]) == current_ngram:
                        count += 1
                
                if count >= self.max_repeats:
                    return " ".join(current_ngram)
        return True

    def reset(self):
        self.buffer.clear()

File: src/core/repetitionDetector.js
/**
 * Detects cyclic repetitions in text by monitoring n-grams in a sliding window buffer.
 */
class RepetitionDetector {
  /**
   * Initializes the RepetitionDetector.
   * @param {number} [n] - The size of the n-gram to monitor for repetitions.
   * @param {number} [maxRepeats] - The maximum number of allowed repetitions before a cycle is detected.
   */
  constructor(n = 20, maxRepeats = 5) {
    this.n = n;
    this.maxRepeats = maxRepeats;
    this.buffer = [];
    this.maxBufferSize = n * maxRepeats * 2;
  }

  /**
   * Analyzes text for cyclic repetitions by tokenizing and updating the sliding window buffer.
   * @param {string} text - The text to analyze for repetitions.
   * @returns {string|boolean} The repeated sequence if a cycle is detected, otherwise true if the text is valid.
   */
  addText(text) {
    if (!text) return true;

    // Simple tokenization based on words
    const tokens = text.match(/\w+/g) || [];

    for (let i = 0; i < tokens.length; i += 1) {
      const token = tokens[i];
      this.buffer.push(token);
      if (this.buffer.length > this.maxBufferSize) {
        this.buffer.shift();
      }

      if (this.buffer.length >= this.n * 2) {
        // Extract current and previous n-grams
        const currentNgram = this.buffer.slice(-this.n).join(" ");
        const previousNgram = this.buffer.slice(-this.n * 2, -this.n).join(" ");

        if (currentNgram === previousNgram) {
          // Count how many times this n-gram appears in the current buffer
          let count = 0;
          for (let j = 0; j <= this.buffer.length - this.n; j += this.n) {
            const segment = this.buffer.slice(j, j + this.n).join(" ");
            if (segment === currentNgram) {
              count += 1;
            }
          }

          if (count >= this.maxRepeats) {
            return currentNgram;
          }
        }
      }
    }

    return true;
  }

  /**
   * Resets the token buffer to clear the repetition history.
   * @returns {void}
   */
  reset() {
    this.buffer = [];
  }
}

export default RepetitionDetector;